internal/aws/k8s/k8sclient/clientset.go

```diff
@@ -0,0 +1,245 @@
+	"strings"
+	"sync"
+	"time"
+
+	"go.uber.org/zap"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+)
+
+const (
+	cacheTTL = 10 * time.Minute
+)
+
+// K8sClientOption is a struct that can be used to change the configuration of passed K8sClient
+// It can be used as an option to the Get(...) function to create a customized K8sClient
+type K8sClientOption struct {
+	name string
+	set  func(*K8sClient)
+}
+
+var mu = &sync.RWMutex{}
+
+// Get returns a singleton instance of k8s client
+// If the intialization fails, it returns nil
+var Get func(logger *zap.Logger, options ...K8sClientOption) *K8sClient
+
+var optionsToK8sClient = map[string]*K8sClient{}
+
+type cacheReflector interface {
+	LastSyncResourceVersion() string
+	Run(<-chan struct{})
+}
+
+type initialSyncChecker interface {
+	// check the initial sync of cache reflector and log the warnMessage if timeout
+	Check(reflector cacheReflector, warnMessage string)
+}
+
+// reflectorSyncChecker implements initialSyncChecker interface
+type reflectorSyncChecker struct {
+	pollInterval time.Duration
+	pollTimeout  time.Duration
+	logger       *zap.Logger
+}
+
+func (r *reflectorSyncChecker) Check(reflector cacheReflector, warnMessage string) {
+	if err := wait.Poll(r.pollInterval, r.pollTimeout, func() (done bool, err error) {
+		return reflector.LastSyncResourceVersion() != "", nil
+	}); err != nil {
+		r.logger.Warn(warnMessage, zap.Error(err))
+	}
+}
+
+// KubeConfigPath provides the option to set the kube config which will be used if the
+// service account that kubernetes gives to pods can't be used
+func KubeConfigPath(kubeConfigPath string) K8sClientOption {
+	return K8sClientOption{
+		name: "kubeConfigPath:" + kubeConfigPath,
+		set: func(kc *K8sClient) {
+			kc.kubeConfigPath = kubeConfigPath
+		},
+	}
+}
+
+// InitSyncPollInterval provides the option to set the init sync poll interval
+// for testing connection to kubernetes api server
+func InitSyncPollInterval(pollInterval time.Duration) K8sClientOption {
+	return K8sClientOption{
+		name: "initSyncPollInterval:" + pollInterval.String(),
+		set: func(kc *K8sClient) {
+			kc.initSyncPollInterval = pollInterval
+		},
+	}
+}
+
+// InitSyncPollInterval provides the option to set the init sync poll timeout
+// for testing connection to kubernetes api server
+func InitSyncPollTimeout(pollTimeout time.Duration) K8sClientOption {
+	return K8sClientOption{
+		name: "initSyncPollTimeout:" + pollTimeout.String(),
+		set: func(kc *K8sClient) {
+			kc.initSyncPollTimeout = pollTimeout
+		},
+	}
+}
+
+func getStringifiedOptions(options ...K8sClientOption) string {
+	opts := make([]string, 0)
+	for _, option := range options {
+		opts = append(opts, option.name)
+	}
+
+	sort.Strings(opts)
+	return strings.Join(opts, "+")
+}
+
+func get(logger *zap.Logger, options ...K8sClientOption) *K8sClient {
+	strOptions := getStringifiedOptions(options...)
+
+	mu.RLock()
```

Comment:
`clientset.get` doesn't seem like it should be called on the hot path, do we need this `RLock` / `Lock` optimization?